op-node/p2p/store/scorebook.go

```diff
@@ -0,0 +1,90 @@
+package store
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/multiformats/go-base32"
+)
+
+type scoreBook struct {
+	ctx   context.Context
+	store ds.Batching
+	cache *lru.Cache[peer.ID, PeerScores]
+	sync.RWMutex
+}
+
+var scoresBase = ds.NewKey("/peers/scores")
+
+type ScoreType string
+
+const (
+	scoreDataV0    = "0"
+	scoreCacheSize = 100
+)
+
+func newScoreBook(ctx context.Context, store ds.Batching) (*scoreBook, error) {
+	cache, err := lru.New[peer.ID, PeerScores](scoreCacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("creating cache: %w", err)
+	}
+	return &scoreBook{
+		ctx:   ctx,
+		store: store,
+		cache: cache,
+	}, nil
+}
+
+func (d *scoreBook) GetPeerScores(id peer.ID) (PeerScores, error) {
+	d.RLock()
+	defer d.RUnlock()
+	return d.getPeerScoresNoLock(id)
+}
+
+func (d *scoreBook) getPeerScoresNoLock(id peer.ID) (PeerScores, error) {
+	scores, ok := d.cache.Get(id)
+	if ok {
+		return scores, nil
+	}
+	data, err := d.store.Get(d.ctx, scoreKey(id))
+	if errors.Is(err, ds.ErrNotFound) {
+		return PeerScores{}, nil
+	} else if err != nil {
+		return PeerScores{}, fmt.Errorf("load scores for peer %v: %w", id, err)
+	}
+	scores, err = deserializeScoresV0(data)
+	if err != nil {
+		return PeerScores{}, fmt.Errorf("invalid score data for peer %v: %w", id, err)
+	}
+	d.cache.Add(id, scores)
+	return scores, nil
+}
+
+func (d *scoreBook) SetGossipScore(id peer.ID, score float64) error {
```

Comment:
I think we should define a Go enum (typed int, iota definition) and change this to `SetScore(id peer.ID, scoreType ScoreType, score float64) error` and then add a switch to write it to the scores struct. Specifically I'd like the gossip-score to be stored with its original parts, so we can see details when we debug why a score is high/low. And this would make adding more non-gossip score types easy.

Reply:
hah, funnily enough that's the way I originally wrote it and then changed to add more type safety.  Can switch it back.